### PR TITLE
feat(sdk): Move `EventTimelineItem.event_id` into `TimelineKey::TransactionId`

### PR DESCRIPTION
**Pull request by Hywan**
To create an event next to the server, the event must be accompanied by a transaction ID. When the server receives the event creation request, it responds with a new event ID (which we will name `event_id_1`). Later on, when a sync is done to the server, the server may respond with the transaction ID and a new event ID (which we will name `event_id_2`).

The transaction ID is used to update the state of the local event. The event ID received from the sync may ideally be the same as the one received by the creation request's response.

Knowing that…

`EventTimelineItem` had a field: `event_id: Option<OwnedEventId>`. It was `Some(event_id)` where `event_id` is the event ID responded by the server when an event was created, i.e. it's `event_id_1`; otherwise it was `None`. This is never `event_id_2`. Why?

Because `EventTimelineItem` has an other field: `key: TimelineKey`, which represents the following states:

* `TimelineKey::TransactionId(OwnedTransactionId)`,
* `TimelineKey::EventId(OwnedEventId)`, where the event ID is `event_id_2`!

So it becomes obvious that `event_id_1` should be part of `TimelineKey`, not `EventTimelineItem`. `TimelineKey` is where this transaction ID and event ID logic is managed.

This patch updates `TimelineyKey::TransactionId` to be defined as:

* `TimelineKey::TransactionId { txn_id: OwnedTransactionId, event_id: Option<OwnedEventId> }`.

Why an `Option`? Because before receiving `event_id_1`, we may still want to create a `TimelineKey`, the API must reflect that state.

So basically, `EventTimelineItem.event_id` is moved inside `TimelineyKey::TransactionId`.

### CODE REVIEW

File: crates/matrix-sdk/src/room/timeline/inner.rs [hunk 1]
```diff
@@ -146,17 +146,26 @@ impl<P: ProfileProvider> TimelineInner<P> {
...
-            if let Some(existing_event_id) = &item.event_id {
-                error!(
-                    ?existing_event_id, new_event_id = ?event_id, ?txn_id,
-                    "Local echo already has an event ID"
+            // It's always a `TimelineKey::TransactionId`
+            if let TimelineKey::TransactionId { txn_id, event_id: txn_event_id } = &item.key {
+                if let Some(existing_event_id) = txn_event_id {
+                    error!(
+                        ?existing_event_id, new_event_id = ?event_id, ?txn_id,
+                        "Local echo already has an event ID"
```
Comment (Hywan) [inline crates/matrix-sdk/src/room/timeline/inner.rs:158]: Should we mention that we are going to override the existing event ID?

File: crates/matrix-sdk/src/room/timeline/event_item.rs [hunk 2]
```diff
@@ -212,14 +214,46 @@ impl EventTimelineItem {
 pub enum TimelineKey {
     /// Transaction ID, for an event that was created locally and hasn't been
     /// acknowledged by the server yet.
-    TransactionId(OwnedTransactionId),
+    TransactionId {
+        /// The transaction ID.
+        txn_id: OwnedTransactionId,
+        /// The event ID received by the server as the new event's response (not
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_item.rs:220]:
```suggestion
        /// The event ID received from the server in the event-sending response (not
```

File: crates/matrix-sdk/src/room/timeline/inner.rs [hunk 3]
```diff
@@ -146,17 +146,26 @@ impl<P: ProfileProvider> TimelineInner<P> {
...
 
+    /// Update the transaction ID by an event ID.
     pub(super) fn add_event_id(&self, txn_id: &TransactionId, event_id: OwnedEventId) {
         let mut lock = self.items.lock_mut();
         if let Some((idx, item)) = find_event_by_txn_id(&lock, txn_id) {
-            if let Some(existing_event_id) = &item.event_id {
-                error!(
-                    ?existing_event_id, new_event_id = ?event_id, ?txn_id,
-                    "Local echo already has an event ID"
+            // It's always a `TimelineKey::TransactionId`
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/inner.rs:153]: Maybe we should write something like this?

```rust
let TimelineKey::TransactionId { txn_id, event_id: txn_event_id } = &item.key else {
    unreachable!("find_event_by_txn_id can only find items with TimelineKey::TransactionId");
};
```

File: crates/matrix-sdk/src/room/timeline/event_item.rs (hunk 2 above)
Comment (Hywan) [reply crates/matrix-sdk/src/room/timeline/event_item.rs:220]: :+1:

File: crates/matrix-sdk/src/room/timeline/inner.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk/src/room/timeline/inner.rs]: :+1:

**Review by jplatte — APPROVED** [review]
"feedback" has no plural. But code looks good ^^

**Comment by Hywan** [comment]
Haha, damned… :-p.
